File: setup/create_circle_wallets.ts
/**
 * Agon — Circle Developer-Controlled Wallet Kurulumu
 *
 * 1. Entity Secret oluşturur ve Circle'a kayıt eder
 * 2. Agent + Evaluator wallet oluşturur (Arc Testnet)
 * 3. Agon .env'e gerekli değerleri yazar
 *
 * Kullanım:
 *   CIRCLE_API_KEY=TEST_API_KEY:...:... node --import=tsx create_circle_wallets.ts
 */

import crypto from "node:crypto";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import {
  registerEntitySecretCiphertext,
  initiateDeveloperControlledWalletsClient,
} from "@circle-fin/developer-controlled-wallets";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT_ENV   = path.join(__dirname, "..", ".env");
const OUTPUT_DIR = path.join(__dirname, "output");

function appendEnv(key: string, value: string) {
  const content = fs.existsSync(ROOT_ENV) ? fs.readFileSync(ROOT_ENV, "utf-8") : "";
  if (content.includes(`${key}=`)) {
    // Güncelle
    fs.writeFileSync(ROOT_ENV, content.replace(new RegExp(`${key}=.*`), `${key}=${value}`));
  } else {
    fs.appendFileSync(ROOT_ENV, `\n${key}=${value}`);
  }
}

async function main() {
  const apiKey = process.env.CIRCLE_API_KEY;
  if (!apiKey) throw new Error("CIRCLE_API_KEY eksik");

  fs.mkdirSync(OUTPUT_DIR, { recursive: true });

  // ── 1. Entity Secret ──────────────────────────────────────────────────────
  console.log("Entity Secret oluşturuluyor ve kayıt ediliyor...");
  const entitySecret = crypto.randomBytes(32).toString("hex");

  await registerEntitySecretCiphertext({
    apiKey,
    entitySecret,
    recoveryFileDownloadPath: OUTPUT_DIR,
  });

  appendEnv("CIRCLE_ENTITY_SECRET", entitySecret);
  appendEnv("CIRCLE_BLOCKCHAIN", "ARC-TESTNET");
  console.log("Entity Secret kayıt edildi ✓");
  console.log(`⚠️  Kurtarma dosyası: ${OUTPUT_DIR}/`);

  // ── 2. SDK client ─────────────────────────────────────────────────────────
  const client = initiateDeveloperControlledWalletsClient({ apiKey, entitySecret });

  // ── 3. Wallet Set ─────────────────────────────────────────────────────────
  console.log("\nWallet Set oluşturuluyor...");
  const walletSet = (await client.createWalletSet({ name: "Agon-WalletSet" })).data?.walletSet;
  if (!walletSet?.id) throw new Error("Wallet Set oluşturulamadı");
  console.log("Wallet Set ID:", walletSet.id);

  // ── 4. Agent + Evaluator wallet ───────────────────────────────────────────
  console.log("\nAgent ve Evaluator wallet oluşturuluyor (Arc Testnet)...");
  const wallets = (await client.createWallets({
    walletSetId: walletSet.id,
    blockchains: ["ARC-TESTNET"],
    count: 2,
    accountType: "EOA",
    metadata: [
      { name: "agon-agent",     refId: "agon-agent"     },
      { name: "agon-evaluator", refId: "agon-evaluator" },
    ],
  })).data?.wallets;

  if (!wallets || wallets.length < 2) throw new Error("Wallet oluşturulamadı");

  const agentWallet     = wallets[0];
  const evaluatorWallet = wallets[1];

  appendEnv("CIRCLE_WALLET_ID",           agentWallet.id);
  appendEnv("CIRCLE_EVALUATOR_WALLET_ID", evaluatorWallet.id);

  fs.writeFileSync(
    path.join(OUTPUT_DIR, "agon-wallets.json"),
    JSON.stringify({ agent: agentWallet, evaluator: evaluatorWallet }, null, 2),
  );

  // ── 5. Özet ──────────────────────────────────────────────────────────────
  console.log("\n" + "=".repeat(55));
  console.log("Kurulum tamamlandı!\n");
  console.log("Agent wallet:");
  console.log("  ID     :", agentWallet.id);
  console.log("  Adres  :", agentWallet.address);
  console.log("\nEvaluator wallet:");
  console.log("  ID     :", evaluatorWallet.id);
  console.log("  Adres  :", evaluatorWallet.address);
  console.log("\n.env güncellendi ✓");
  console.log("=".repeat(55));
  console.log("\nSonraki adımlar:");
  console.log("1. Her iki adrese Arc Testnet faucet'ten USDC gönder:");
  console.log("   https://faucet.circle.com → Arc Testnet");
  console.log(`   Agent    : ${agentWallet.address}`);
  console.log(`   Evaluator: ${evaluatorWallet.address}`);
  console.log("\n2. Agent adresini IdentityRegistry'e kayıt et:");
  console.log(`   cast send 0x8004A818BFB912233c491871b3d84c89A494BD9e \\`);
  console.log(`     "register(string)(uint256)" "ipfs://QmAgentBountyProtocol" \\`);
  console.log(`     --from ${agentWallet.address} --rpc-url https://rpc.testnet.arc.network`);
}

main().catch((err) => {
  console.error("Hata:", err.message || err);
  process.exit(1);
});

File: setup/create_circle_wallets.py
"""
Circle Developer-Controlled Wallet kurulum scripti.

Agon için iki wallet oluşturur:
  1. Agent wallet   (takeTask + submitResult)
  2. Evaluator wallet (approveResult + rejectResult)

Çalıştırmadan önce .env'de şunlar olmalı:
  CIRCLE_API_KEY
  CIRCLE_ENTITY_SECRET
  CIRCLE_BLOCKCHAIN   (örn: "ARC-TESTNET")

Çıktı: .env'e eklenecek CIRCLE_WALLET_ID ve CIRCLE_EVALUATOR_WALLET_ID değerleri.

Kullanım:
  pip install requests cryptography
  python setup/create_circle_wallets.py
"""

import os
import base64
import requests
from uuid import uuid4
from dotenv import load_dotenv
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import padding

load_dotenv()

API_KEY       = os.environ["CIRCLE_API_KEY"]
ENTITY_SECRET = os.environ["CIRCLE_ENTITY_SECRET"]
BLOCKCHAIN    = os.environ["CIRCLE_BLOCKCHAIN"]  # örn: "ARC-TESTNET"

BASE = "https://api.circle.com/v1/w3s"


def get_ciphertext() -> str:
    resp = requests.get(
        f"{BASE}/config/entity/publicKey",
        headers={"Authorization": f"Bearer {API_KEY}"},
        timeout=10,
    )
    resp.raise_for_status()
    public_key_pem = resp.json()["data"]["publicKey"]

    secret_bytes = bytes.fromhex(ENTITY_SECRET)
    public_key   = serialization.load_pem_public_key(public_key_pem.encode())
    encrypted    = public_key.encrypt(
        secret_bytes,
        padding.OAEP(
            mgf=padding.MGF1(algorithm=hashes.SHA256()),
            algorithm=hashes.SHA256(),
            label=None,
        ),
    )
    return base64.b64encode(encrypted).decode()


def create_wallet_set(name: str) -> str:
    resp = requests.post(
        f"{BASE}/developer/walletSets",
        headers={
            "Authorization": f"Bearer {API_KEY}",
            "Content-Type":  "application/json",
        },
        json={
            "idempotencyKey":          str(uuid4()),
            "entitySecretCiphertext":  get_ciphertext(),
            "name":                    name,
        },
        timeout=15,
    )
    resp.raise_for_status()
    wallet_set_id = resp.json()["data"]["walletSet"]["id"]
    print(f"  Wallet set oluşturuldu: {wallet_set_id}")
    return wallet_set_id


def create_wallet(wallet_set_id: str, name: str) -> dict:
    resp = requests.post(
        f"{BASE}/developer/wallets",
        headers={
            "Authorization": f"Bearer {API_KEY}",
            "Content-Type":  "application/json",
        },
        json={
            "idempotencyKey":          str(uuid4()),
            "entitySecretCiphertext":  get_ciphertext(),
            "walletSetId":             wallet_set_id,
            "blockchains":             [BLOCKCHAIN],
            "count":                   1,
            "accountType":             "EOA",
            "metadata":                [{"name": name, "refId": name}],
        },
        timeout=15,
    )
    resp.raise_for_status()
    wallet = resp.json()["data"]["wallets"][0]
    return wallet


def main():
    print("=" * 50)
    print("Agon — Circle Wallet Kurulumu")
    print("=" * 50)
    print(f"Blockchain: {BLOCKCHAIN}\n")

    print("[1/2] Agent wallet set oluşturuluyor...")
    agent_set_id = create_wallet_set("Agon-Agent-WalletSet")

    print("[2/2] Wallet'lar oluşturuluyor...")
    agent_wallet     = create_wallet(agent_set_id, "agon-agent")
    evaluator_wallet = create_wallet(agent_set_id, "agon-evaluator")

    agent_addr     = agent_wallet["address"]
    evaluator_addr = evaluator_wallet["address"]
    agent_id       = agent_wallet["id"]
    evaluator_id   = evaluator_wallet["id"]

    print("\n" + "=" * 50)
    print("Kurulum tamamlandı. .env'e ekle:\n")
    print(f"CIRCLE_WALLET_ID={agent_id}")
    print(f"CIRCLE_EVALUATOR_WALLET_ID={evaluator_id}")
    print()
    print("Wallet adresleri (IdentityRegistry'e kayıt için):")
    print(f"  Agent     : {agent_addr}")
    print(f"  Evaluator : {evaluator_addr}")
    print()
    print("ÖNEMLI: Bu adresleri IdentityRegistry'e kayıt ettirmen gerekiyor.")
    print("Sonra her wallet'a Arc Testnet'ten USDC gönder (stake + gas için).")
    print("=" * 50)


if __name__ == "__main__":
    main()
